crates/ark/src/analysis/parse_boundaries.rs

```diff
@@ -0,0 +1,375 @@
+
+    Ok(boundaries)
+}
+
+fn merge_overlapping<T>(ranges: Vec<std::ops::Range<T>>) -> Vec<std::ops::Range<T>>
+where
+    T: PartialOrd + Ord + Copy,
+{
+    let merge = |mut merged: Vec<std::ops::Range<T>>, range: std::ops::Range<T>| {
+        if let Some(last) = merged.last_mut() {
+            // if range.start <= last.end {
+            if last.contains(&range.start) {
+                // Overlap, merge with last range
+                last.end = last.end.max(range.end);
+            } else {
+                // No overlap, push a new range
+                merged.push(range);
+            }
+        } else {
+            // First element, just add it
+            merged.push(range);
+        }
+
+        merged
+    };
+
+    ranges.into_iter().fold(vec![], merge)
+}
+
+// Fill any gaps with one-liner complete expressions. This applies
+// to empty lines, with or without spaces and tabs, and possibly
+// with a trailing comment.
+fn fill_gaps(mut boundaries: ParseBoundaries, n_lines: usize) -> ParseBoundaries {
+    let mut filled = vec![];
+    let ranges = boundaries.complete;
+
+    let mut last_line = 0;
+
+    let range_from = |start| std::ops::Range {
+        start,
+        end: start + 1,
+    };
+
+    // Fill leading whitespace with empty input ranges
+    if let Some(first) = ranges
+        .get(0)
+        .or(boundaries.incomplete.as_ref())
+        .or(boundaries.error.as_ref())
+    {
+        for start in 0..first.start {
+            let range = range_from(start);
+            last_line = range.end;
+            filled.push(range)
+        }
+    }
+
+    // Fill gaps between complete expressions
+    {
+        let iter = ranges.into_iter();
+        for range in iter {
+            // We found a gap, fill ranges for lines in that gap
+            if !range.contains(&last_line) {
+                for start in last_line..range.start {
+                    filled.push(range_from(start))
+                }
+            }
+
+            last_line = range.end;
+            filled.push(range);
+        }
+    }
+
+    // Fill trailing whitespace (between complete and incomplete|error\eof)
+    let last_boundary = filled.last().map(|r| r.end).unwrap_or(0);
+    let next_boundary = boundaries
+        .incomplete
+        .as_ref()
+        .or(boundaries.error.as_ref())
+        .map(|r| r.start)
+        .unwrap_or(n_lines);
+
+    for start in last_boundary..next_boundary {
+        filled.push(range_from(start))
+    }
+
+    boundaries.complete = filled;
+    boundaries
+}
+
+#[cfg(test)]
+mod tests {
+    use harp::assert_match;
+
+    use crate::analysis::parse_boundaries::*;
+    use crate::test::r_test;
+
+    #[test]
+    fn test_parse_boundaries_complete() {
+        r_test(|| {
+            let boundaries = parse_boundaries("foo").unwrap();
+            #[rustfmt::skip]
+            assert_eq!(boundaries.complete, vec![
+                std::ops::Range { start: 0, end: 1 },
```

Comment:
A custom type would also let us add a `new()` method for something like `LineRange::new(0, 1)`

Reply:
Good point though that's arguably straying away from the standard style.